llvm/lib/CAS/BuiltinCAS.h

```diff
@@ -0,0 +1,200 @@
+//===- BuiltinCAS.h ---------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_CAS_BUILTINCAS_H
+#define LLVM_LIB_CAS_BUILTINCAS_H
+
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/CAS/CASDB.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/SHA1.h"
+#include <cstddef>
+
+namespace llvm {
+namespace cas {
+namespace builtin {
+
+/// Dereference the inner value in \p E, adding an Optional to the outside.
+/// Useful for stripping an inner Optional in return chaining.
+///
+/// \code
+/// Expected<Optional<SomeType>> f1(...);
+///
+/// Expected<SomeType> f2(...) {
+///   if (Optional<Expected<NoneType>> E = dereferenceValue(f1()))
+///     return std::move(*E);
+///
+///   // Deal with None...
+/// }
+/// \endcode
+template <class T>
+inline Optional<Expected<T>> dereferenceValue(Expected<Optional<T>> E) {
```

Comment:
Why are these in llvm::cas::builtin? These look like they are general purpose.  Maybe they should be members of Expected?

I find the name "dereference" unclear, since it does not in fact dereference the optional.  In Rust this operation is called "transpose", which seems at least a bit better to me: https://doc.rust-lang.org/std/result/enum.Result.html#method.transpose

Reply:
> Why are these in llvm::cas::builtin? These look like they are general purpose. Maybe they should be members of Expected?

Only because they were in BuiltinCAS.cpp before (still in a private header, too). Totally general purpose, just haven't moved them somewhere general yet :/.

`transpose` seems nice!

Reply:
Thoughts on the naming of the second function?
```
template <class T>
inline Expected<T> dereferenceValue(Expected<Optional<T>> E,
                                    function_ref<Error()> OnNone) {
  if (Optional<Expected<T>> MaybeExpected = transpose(std::move(E)))
    return std::move(*MaybeExpected);
  return OnNone();
}
```

Reply:
For now I've left it as is. Renamed the one you commented on to `transpose` in bc83886e2b1da963a66ab30138b66d693c353b5a.

(Haven't moved it yet.)

Reply:
May not do this today so I added FIXMEs.